# <center> Data Cleaning of the Orders and Line Items Tables </center>

# Importing the packages

In [1]:
import pandas as pd
import numpy as np

# Importing the Tables

In [2]:
orders_df = pd.read_excel('Printify_Data_Analyst_homework.xlsx', sheet_name = 'Orders')
orders_df.head()

MERCHANT_ID     ORDER_ID  SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
0      7143578  2013244.178  2013244                 US                WA   
1      5888534  1358429.695  1358429                 US                MN   
2      6484224   939390.193   939390                 US                AK   
3      7251033  1856107.400  1856107                 US                PA   
4      7241488  1843904.400  1843904                 US                CO   

             ORDER_DT        FULFILLED_DT  REPRINT_FLAG  \
0 2020-10-02 19:30:15 2020-10-20 07:53:06           NaN   
1 2020-11-21 05:19:03 2020-11-23 00:23:32           NaN   
2 2020-10-25 01:39:08 2020-10-26 20:30:53           NaN   
3 2020-10-08 16:35:21 2020-10-10 15:02:47           NaN   
4 2020-11-17 19:30:10 2020-11-18 03:45:25           NaN   

   SALES_CHANNEL_TYPE_ID  TOTAL_COST  TOTAL_SHIPPING MERCHANT_REGISTERED_DT  \
0                      9        2902            2200    2020-05-19 15:22:57   
1                      4        1149             924    2018-02-08 07:34:02   
2                      1         977             400    2019-04-09 20:02:46   
3                      1         661             800    2020-07-20 17:04:41   
4                      2         719             424    2020-07-15 18:47:51   

   SUB_IS_ACTIVE_FLAG            SUB_PLAN SHIPMENT_CARRIER  \
0                True  business_account_3              UPS   
1                True  business_account_3             USPS   
2                True  business_account_3             USPS   
3                True  business_account_3              UPS   
4                True  business_account_3             USPS   

  SHIPMENT_DELIVERD_DT  
0  2020-10-08 21:07:29  
1  2020-11-28 21:50:56  
2  2020-11-02 23:37:54  
3  2020-10-15 18:08:37  
4  2020-11-24 00:44:56

In [3]:
line_items_df = pd.read_excel('Printify_Data_Analyst_homework.xlsx', sheet_name = 'Line items')
line_items_df.head()

ORDER_ID  PRINT_PROVIDER_ID  PRODUCT_BRAND  PRODUCT_TYPE  \
0  1.750015e+06                 30  Generic brand           Mug   
1  1.750015e+06                 30  Generic brand           Mug   
2  1.225155e+06                 16         Gildan  Kids clothes   
3  1.828891e+06                 39   Bella+Canvas       T-Shirt   
4  1.847848e+06                 10  Generic brand   Accessories   

          ITEM_STATUS  QUANTITY  
0             on-hold         3  
1             on-hold         2  
2  shipment_delivered         1  
3  shipment_delivered         1  
4  shipment_delivered         1

# Working with the line_items_df
Since it has fewer fields, I'll start the data cleaning with this one

In [4]:
# First, I want to know a bit more about this df
line_items_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 57583 entries, 0 to 57582
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   ORDER_ID           57583 non-null  float64
 1   PRINT_PROVIDER_ID  57583 non-null  int64  
 2   PRODUCT_BRAND      57583 non-null  object 
 3   PRODUCT_TYPE       57583 non-null  object 
 4   ITEM_STATUS        57583 non-null  object 
 5   QUANTITY           57583 non-null  int64  
dtypes: float64(1), int64(2), object(3)
memory usage: 2.6+ MB


In [5]:
# From my experience is better to save IDs, or columns that can be used as Primary Keys, as strings, so I'll do the conversion:
line_items_df['ORDER_ID'] = line_items_df['ORDER_ID'].astype(str)
line_items_df['PRINT_PROVIDER_ID'] = line_items_df['PRINT_PROVIDER_ID'].astype(str)

line_items_df.dtypes # verifying the columns' data types

ORDER_ID             object
PRINT_PROVIDER_ID    object
PRODUCT_BRAND        object
PRODUCT_TYPE         object
ITEM_STATUS          object
QUANTITY              int64
dtype: object

## Working on the ORDER_ID column

In [6]:
# Looking for Nulls
sum(line_items_df['ORDER_ID'].isna())

0

There are no Nulls, as should be expected from a column that stores IDs and can work as a Foreign Key.

In [7]:
# Getting a glance of the data in this column
line_items_df['ORDER_ID'].value_counts()

1750015.16551    15651
1750015.14603     7335
1750015.14371     5145
1684398.1252      1815
1750015.15364     1218
                 ...  
2231884.334          1
1657988.1181         1
1499068.587          1
130592.5851          1
1991060.13           1
Name: ORDER_ID, Length: 13466, dtype: int64

By using the <code>.value_counts()</code> method I can see the order_ids and how many lines of items each have.

In [8]:
# Based on the previous result, I'd like to know to number of unique values for the ORDER_ID column
line_items_df['ORDER_ID'].nunique()

13466

## Working on the PRINT_PROVIDER_ID column
I'll do the same as in the previous column.

In [9]:
# Looking for Nulls
sum(line_items_df['PRINT_PROVIDER_ID'].isna())

0

In [10]:
# Taking a glance at the data in this column and checking if the PRINT_PROVIDER_ID values have blanks or typos
line_items_df['PRINT_PROVIDER_ID'].value_counts().sort_index()

1      1793
10     2070
14       83
16     3471
2       648
23      200
25      476
26      192
27      107
28       84
29     5590
3      2593
30    35837
33      140
34       40
36       16
37       38
38       29
39     1369
40       26
41      666
42      239
43      172
44      154
45      517
46        3
47       16
48      267
49        1
51       69
52      140
54       17
59      243
6       148
65       26
7        16
70       14
8         1
9        72
Name: PRINT_PROVIDER_ID, dtype: int64

In [11]:
# Based on the previous result, I'd like to know to number of unique values for the PRINT_PROVIDER_ID column
line_items_df['PRINT_PROVIDER_ID'].nunique()

39

## Working on the PRODUCT_BRAND column
I'll do the same as in the previous column.

In [12]:
# Looking for Nulls
sum(line_items_df['PRODUCT_BRAND'].isna())

0

In [13]:
# Taking a glance at the data in this column and checking if the PRODUCT_BRAND values have blanks or typos
line_items_df['PRODUCT_BRAND'].value_counts().sort_index()

AS Colour                         10
AWDIS                             91
Alstyle                            1
Anvil                             18
B&C                               44
Bella+Canvas                    3981
Case Mate                         35
Champion                          71
Delta                            127
District                           6
Fruit Of The Loom                 31
Generic brand                  44455
Gildan                          7439
Hanes                              8
Independent Trading Company        1
LAT Apparel                       23
Lane Seven                       135
Next Level                       722
Premium Cases                    198
Rabbit Skins                      82
SG Clothing                        1
Sport-Tek                         77
Stanley Stella                    27
Name: PRODUCT_BRAND, dtype: int64

In [14]:
# Based on the previous result, I'd like to know to number of unique values for the PRODUCT_BRAND column
line_items_df['PRODUCT_BRAND'].nunique()

23

## Working on the PRODUCT_TYPE column

In [15]:
# Looking for Nulls
sum(line_items_df['PRODUCT_TYPE'].isna())

0

In [16]:
# Taking a glance at the data in this column and checking if the PRODUCT_TYPE values have blanks or typos
line_items_df['PRODUCT_TYPE'].value_counts().sort_index()

Accessories         1224
All Over Prints      601
Bags                 386
Calendar               1
Canvas               182
Hats                 208
Home Decor          3173
Hoodie              1644
Kids clothes         341
Laptop Sleeve         14
Long-sleeve          350
Mug                37413
Paper products       680
Phone Case           233
Poster               541
Puzzle                18
Shoes                 19
Sweatshirt          3000
T-Shirt             7317
Tank Top             148
Trousers               8
V-neck                82
Name: PRODUCT_TYPE, dtype: int64

In [17]:
# Based on the previous result, I'd like to know to number of unique values for the PRODUCT_TYPE column
line_items_df['PRODUCT_TYPE'].nunique()

22

## Working on the ITEM_STATUS column

In [18]:
# Looking for Nulls
sum(line_items_df['ITEM_STATUS'].isna())

0

In [19]:
# Taking a glance at the data in this column and checking if the ITEM_STATUS values have blanks or typos
line_items_df['ITEM_STATUS'].value_counts().sort_index()

canceled                          1168
fulfilled                         1793
in-production                       32
on-hold                          39071
shipment_available_for_pickup       13
shipment_cancelled                   1
shipment_delivered               13299
shipment_error                       2
shipment_failure                    32
shipment_in_transit                795
shipment_out_for_delivery         1276
shipment_pre_transit                87
shipment_return_to_sender           14
Name: ITEM_STATUS, dtype: int64

In [20]:
# Based on the previous result, I'd like to know to number of unique values for the ITEM_STATUS column
line_items_df['ITEM_STATUS'].nunique()

13

## Working on the QUANTITY column

In [21]:
# Looking for Nulls
sum(line_items_df['QUANTITY'].isna())

0

In [22]:
# Getting a glance of the data in this column
line_items_df['QUANTITY'].value_counts()

1      54052
2       3052
3        340
4         67
5         28
6         14
10         9
7          7
8          4
15         3
25         1
30         1
16         1
40         1
12         1
11         1
100        1
Name: QUANTITY, dtype: int64

For the moment, I think I did all the cleaning that I needed to do for this df. However, based on what I find in the **orders_df** I may return to this df and change some values. So let's see how the data looks with the changes made so far:

In [23]:
line_items_df.head()

ORDER_ID PRINT_PROVIDER_ID  PRODUCT_BRAND  PRODUCT_TYPE  \
0  1750015.15208                30  Generic brand           Mug   
1  1750015.12357                30  Generic brand           Mug   
2  1225155.10081                16         Gildan  Kids clothes   
3    1828890.707                39   Bella+Canvas       T-Shirt   
4     1847847.55                10  Generic brand   Accessories   

          ITEM_STATUS  QUANTITY  
0             on-hold         3  
1             on-hold         2  
2  shipment_delivered         1  
3  shipment_delivered         1  
4  shipment_delivered         1

# Working with the orders_df

In [24]:
# Getting a quick overview of the table
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13506 entries, 0 to 13505
Data columns (total 16 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   MERCHANT_ID             13506 non-null  int64         
 1   ORDER_ID                13506 non-null  float64       
 2   SHOP_ID                 13506 non-null  int64         
 3   ADDRESS_TO_COUNTRY      13505 non-null  object        
 4   ADDRESS_TO_REGION       12930 non-null  object        
 5   ORDER_DT                13506 non-null  datetime64[ns]
 6   FULFILLED_DT            12637 non-null  datetime64[ns]
 7   REPRINT_FLAG            103 non-null    float64       
 8   SALES_CHANNEL_TYPE_ID   13506 non-null  int64         
 9   TOTAL_COST              13506 non-null  int64         
 10  TOTAL_SHIPPING          13506 non-null  int64         
 11  MERCHANT_REGISTERED_DT  13506 non-null  datetime64[ns]
 12  SUB_IS_ACTIVE_FLAG      13506 non-null  bool  

In [25]:
#Since all the IDs are either floats or ints, I'll convert them to strings:

orders_df['MERCHANT_ID'] = orders_df['MERCHANT_ID'].astype(str)
orders_df['ORDER_ID'] = orders_df['ORDER_ID'].astype(str)
orders_df['SHOP_ID'] = orders_df['SHOP_ID'].astype(str)
orders_df['SALES_CHANNEL_TYPE_ID'] = orders_df['SALES_CHANNEL_TYPE_ID'].astype(str)

orders_df.dtypes

MERCHANT_ID                       object
ORDER_ID                          object
SHOP_ID                           object
ADDRESS_TO_COUNTRY                object
ADDRESS_TO_REGION                 object
ORDER_DT                  datetime64[ns]
FULFILLED_DT              datetime64[ns]
REPRINT_FLAG                     float64
SALES_CHANNEL_TYPE_ID             object
TOTAL_COST                         int64
TOTAL_SHIPPING                     int64
MERCHANT_REGISTERED_DT    datetime64[ns]
SUB_IS_ACTIVE_FLAG                  bool
SUB_PLAN                          object
SHIPMENT_CARRIER                  object
SHIPMENT_DELIVERD_DT      datetime64[ns]
dtype: object

For now, it looks like I don't need to change more of the table data types, so I'll start working on each individual column.

## Working on the MERCHANT_ID column

In [26]:
# Looking for Nulls in the MERCHANT_ID column
sum(orders_df['MERCHANT_ID'].isna())

0

In [27]:
# Getting a glance of the data in this column
orders_df['MERCHANT_ID'].value_counts()

7019548    150
6433350    106
7084134     99
6795017     81
5755049     81
          ... 
7319108      1
6980095      1
7370726      1
6793352      1
7144356      1
Name: MERCHANT_ID, Length: 4660, dtype: int64

In [28]:
# Based on the previous result, I'd like to know to number of unique values for this column
orders_df['MERCHANT_ID'].nunique()

4660

## Working on the ORDER_ID column

In [29]:
# Looking for any blanks in the ORDER_ID column
sum(orders_df['ORDER_ID'].isna())

0

In [30]:
# Getting a glance of the data in this column
orders_df['ORDER_ID'].value_counts()

516066.48428    2
752154.3        2
2119857.64      2
2232640.7       2
1996849.47      2
               ..
1818327.404     1
1414822.7978    1
1837424.2136    1
2183778.34      1
2216957.28      1
Name: ORDER_ID, Length: 13495, dtype: int64

**Assumption 1:** This table should have the ORDER_ID column as a Primary Key because it is detailing all the orders placed. Therefore, the values should be unique for each record.

However, some ORDER_IDs are being duplicated. I need to dive deeper into these records.

In [31]:
# Finding out the number of order_ids that are duplicated
len(orders_df['ORDER_ID'].value_counts()[orders_df['ORDER_ID'].value_counts() > 1])

11

In [32]:
# Getting the order_ids that are duplicated
orders_df['ORDER_ID'].value_counts()[orders_df['ORDER_ID'].value_counts() > 1]

516066.48428    2
752154.3        2
2119857.64      2
2232640.7       2
1996849.47      2
2229889.102     2
2078441.1       2
2015444.3       2
1820693.162     2
363596.227      2
1555863.282     2
Name: ORDER_ID, dtype: int64

Since it's only 11 Order_IDs that are duplicated, I'll check them one by one.

### ORDER_ID 516066.48428

In [33]:
# Checking the content of the rows
duplicated_order_ids = list(orders_df['ORDER_ID'].value_counts()[orders_df['ORDER_ID'].value_counts() > 1].index)
orders_df[orders_df['ORDER_ID'] == duplicated_order_ids[0]]

MERCHANT_ID      ORDER_ID SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
5436     6118692  516066.48428  516066                 US                MD   
5437     6118692  516066.48428  516066                 US                MD   

                ORDER_DT        FULFILLED_DT  REPRINT_FLAG  \
5436 2020-10-11 04:30:55 2020-10-13 01:11:15           NaN   
5437 2020-10-11 04:30:55 2020-10-13 01:11:15           NaN   

     SALES_CHANNEL_TYPE_ID  TOTAL_COST  TOTAL_SHIPPING MERCHANT_REGISTERED_DT  \
5436                     1        1979             850    2018-09-13 12:23:50   
5437                     1        1979             850    2018-09-13 12:23:50   

      SUB_IS_ACTIVE_FLAG            SUB_PLAN      SHIPMENT_CARRIER  \
5436                True  business_account_3                  USPS   
5437                True  business_account_3  UPS_MAIL_INNOVATIONS   

     SHIPMENT_DELIVERD_DT  
5436  2020-10-16 19:44:40  
5437  2020-10-19 19:21:56

It looks like both rows belong to the same order, which would make it duplicated. The only differences bewteen these 2 rows are in the Shipment_Carrier column and the Shipment_Deliverd_DT. Other than that, the rest of the values are the same. So maybe the ID being duplicated has to do with the fact that it used different carriers to deliver the goods in different times. I'd like to check the values for this order in the **line_items_df** though.

In [34]:
# Checking the items that belong to this order
line_items_df[line_items_df['ORDER_ID'] == duplicated_order_ids[0]]

ORDER_ID PRINT_PROVIDER_ID  PRODUCT_BRAND PRODUCT_TYPE  \
10807  516066.48428                25     Next Level      T-Shirt   
10808  516066.48428                25     Next Level      T-Shirt   
10809  516066.48428                16  Generic brand         Hats   
10810  516066.48428                16  Generic brand         Hats   

              ITEM_STATUS  QUANTITY  
10807  shipment_delivered         1  
10808  shipment_delivered         1  
10809  shipment_delivered         1  
10810  shipment_delivered         1

The problem with this ID being duplicated is that if I want to apply aggregate functions on the Total_Cost or Total_Shipping, the result would not be accurate. I could drop one of these rows, but by doing so I would lose information on the Shipment_Carrier or the Shipment_Deliverd_DT data. I think a fair solution would be to create a new column that would indicate if an ID was duplicated or not, and then use that column to filter out those dupplicates when using an aggregate function. But before doing this, I need to check the rest of the IDs to see if they are also duplicated. This is a **duplicated record**.

**Assumption 2:** For a record in the ORDER_ID column in orders_df to be considered duplicated, the only values that don't have to match are regarding the Shipment_Carrier and/or the Shipment_Deliverd_DT columns. If the ORDER_DT, TOTAL_COST or TOTAL_SHIPPING don't match, then the Order_Ids will not be considered as duplicated but as different IDs, and therefore will be modified.

### ORDER_ID 752154.3

In [35]:
# Checking the content of the rows
orders_df[orders_df['ORDER_ID'] == duplicated_order_ids[1]]

MERCHANT_ID  ORDER_ID SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
13     6314037  752154.3  752154                 US                WA   
14     6314037  752154.3  752154                 US                WA   

              ORDER_DT        FULFILLED_DT  REPRINT_FLAG  \
13 2020-10-03 13:05:53 2020-10-07 20:20:59           NaN   
14 2020-10-03 13:05:53 2020-10-07 20:20:59           NaN   

   SALES_CHANNEL_TYPE_ID  TOTAL_COST  TOTAL_SHIPPING MERCHANT_REGISTERED_DT  \
13                     2       14732            3884    2019-01-25 17:38:56   
14                     2       14732            3884    2019-01-25 17:38:56   

    SUB_IS_ACTIVE_FLAG            SUB_PLAN SHIPMENT_CARRIER  \
13               False  business_account_3             USPS   
14               False  business_account_3              NaN   

   SHIPMENT_DELIVERD_DT  
13  2020-10-09 21:38:49  
14                  NaT

In [36]:
# Checking the items that belong to this order
line_items_df[line_items_df['ORDER_ID'] == duplicated_order_ids[1]]

ORDER_ID PRINT_PROVIDER_ID PRODUCT_BRAND PRODUCT_TYPE  \
25052  752154.3                 3  Bella+Canvas  Long-sleeve   
25053  752154.3                 3  Bella+Canvas  Long-sleeve   
25054  752154.3                 3  Bella+Canvas  Long-sleeve   
25055  752154.3                 3  Bella+Canvas  Long-sleeve   
25056  752154.3                16    Next Level      T-Shirt   
25057  752154.3                16    Next Level      T-Shirt   
25058  752154.3                 3  Bella+Canvas  Long-sleeve   
25059  752154.3                 3  Bella+Canvas  Long-sleeve   
25060  752154.3                45  Bella+Canvas      T-Shirt   
25061  752154.3                45  Bella+Canvas      T-Shirt   
25062  752154.3                45  Bella+Canvas      T-Shirt   
25063  752154.3                45  Bella+Canvas      T-Shirt   
25064  752154.3                16    Next Level      T-Shirt   
25065  752154.3                16    Next Level      T-Shirt   
25066  752154.3                 6    Lane Seven   Sweatshirt   
25067  752154.3                 6    Lane Seven   Sweatshirt   
25068  752154.3                 6    Lane Seven   Sweatshirt   
25069  752154.3                 6    Lane Seven   Sweatshirt   

              ITEM_STATUS  QUANTITY  
25052  shipment_delivered         1  
25053  shipment_delivered         1  
25054  shipment_delivered         1  
25055  shipment_delivered         1  
25056  shipment_delivered         1  
25057  shipment_delivered         1  
25058  shipment_delivered         1  
25059  shipment_delivered         1  
25060           fulfilled         1  
25061           fulfilled         1  
25062           fulfilled         1  
25063           fulfilled         1  
25064  shipment_delivered         1  
25065  shipment_delivered         1  
25066           fulfilled         1  
25067           fulfilled         1  
25068           fulfilled         1  
25069           fulfilled         1

This is also a **duplicated record**. All the values are the same except for shipment. Upon seeing the line_items table I see that some were delivered while some were just printed but not shipped, which makes sense considering that one of the duplicated rows doesn't display the shipment carrier nor date info.

### ORDER_ID 2119857.64

In [37]:
# Checking the content of the rows
orders_df[orders_df['ORDER_ID'] == duplicated_order_ids[2]]

MERCHANT_ID    ORDER_ID  SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
7321      7476675  2119857.64  2119857                 US                OH   
12579     7476675  2119857.64  2119857                 US                TX   

                 ORDER_DT        FULFILLED_DT  REPRINT_FLAG  \
7321  2020-11-25 07:39:40 2020-11-26 20:37:35           NaN   
12579 2020-10-28 18:46:53 2020-10-29 18:58:29           NaN   

      SALES_CHANNEL_TYPE_ID  TOTAL_COST  TOTAL_SHIPPING  \
7321                      4         609             424   
12579                     4         739             400   

      MERCHANT_REGISTERED_DT  SUB_IS_ACTIVE_FLAG SUB_PLAN SHIPMENT_CARRIER  \
7321     2020-10-01 08:44:55               False      NaN             USPS   
12579    2020-10-01 08:44:55               False      NaN             USPS   

      SHIPMENT_DELIVERD_DT  
7321   2020-12-08 16:33:58  
12579  2020-11-02 19:33:48

This Order_ID doesn't seem to be duplicated, many fields values don't match, especially the Total_Cost and Order_DT, which leads me to believe that these records share the same Order_Id by mistake. Another indicator for this is the index. In the previous order_ids, the index were continuous and here there is a gap from index 7321 to 12579. So they definitely are **different ORDER_IDs**.

In [38]:
# Checking the items that belong to this order
line_items_df[line_items_df['ORDER_ID'] == duplicated_order_ids[2]]

ORDER_ID PRINT_PROVIDER_ID PRODUCT_BRAND PRODUCT_TYPE  \
569    2119857.64                 3  Bella+Canvas      T-Shirt   
56154  2119857.64                29        Gildan      T-Shirt   

              ITEM_STATUS  QUANTITY  
569    shipment_delivered         1  
56154  shipment_delivered         1

I can see by checking the index of these rows that the items are not continuous, just like the order_ids in the orders_df. Because of this, and the evidence I noted in the previous cell, I'm convinced this case refers to different Order_Ids and not a duplicated one. I'll change the Order_ID value for the row with the highest index number by adding an 'A' at the end, both in the orders_df as well as in the line_items_df.

In [13]:
# Modifying the order_id values
orders_df.loc[12579,'ORDER_ID'] = '2119857.64A'
line_items_df.loc[56154,'ORDER_ID'] = '2119857.64A'

print(orders_df.loc[12579,'ORDER_ID'])
print(line_items_df.loc[56154,'ORDER_ID'])

2119857.64A
2119857.64A


In [14]:
# Verifying that this order_id is no longer duplicated in the orders_df
orders_df[orders_df['ORDER_ID'] == duplicated_order_ids[2]]

NameError: name 'duplicated_order_ids' is not defined

In [41]:
# Verifying that this order_id is no longer duplicated in the line_items_df
line_items_df[line_items_df['ORDER_ID'] == duplicated_order_ids[2]]

ORDER_ID PRINT_PROVIDER_ID PRODUCT_BRAND PRODUCT_TYPE  \
569  2119857.64                 3  Bella+Canvas      T-Shirt   

            ITEM_STATUS  QUANTITY  
569  shipment_delivered         1

### ORDER_ID 2232640.7

In [42]:
# Checking the content of the rows
orders_df[orders_df['ORDER_ID'] == duplicated_order_ids[3]]

MERCHANT_ID   ORDER_ID  SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
10478     7572140  2232640.7  2232640                 US                IN   
10479     7572140  2232640.7  2232640                 US                IN   

                 ORDER_DT FULFILLED_DT  REPRINT_FLAG SALES_CHANNEL_TYPE_ID  \
10478 2020-12-07 18:36:13          NaT           NaN                     2   
10479 2020-11-05 18:13:25          NaT           NaN                     2   

       TOTAL_COST  TOTAL_SHIPPING MERCHANT_REGISTERED_DT  SUB_IS_ACTIVE_FLAG  \
10478           0               0    2020-11-02 15:41:56                True   
10479           0               0    2020-11-02 15:41:56                True   

                 SUB_PLAN SHIPMENT_CARRIER SHIPMENT_DELIVERD_DT  
10478  business_account_3              NaN                  NaT  
10479  business_account_3              NaN                  NaT

The ORDER_DT values don't match, therefore these records belong to **different ORDER_IDs**. I'll check the line_itmes_df regarding this ORDER_ID to see why the costs are 0.

In [43]:
# Checking the items that belong to this order
line_items_df[line_items_df['ORDER_ID'] == duplicated_order_ids[3]]

ORDER_ID PRINT_PROVIDER_ID  PRODUCT_BRAND PRODUCT_TYPE ITEM_STATUS  \
8895   2232640.7                29         Gildan       Hoodie    canceled   
8896   2232640.7                51  Generic brand  Accessories    canceled   
8897   2232640.7                51  Generic brand  Accessories    canceled   
51662  2232640.7                29   Bella+Canvas      T-Shirt    canceled   
51663  2232640.7                29   Bella+Canvas      T-Shirt    canceled   

       QUANTITY  
8895          1  
8896          1  
8897          2  
51662         1  
51663         1

The items were cancelled, so that would explain why the TOTAL_COST column id orders_df was 0. Since there are 3 items that are continuous, they belong to the first order_id, and the remaining 2 belong to the second order_id. I'll modify the second order_id the same way I previously did.

In [44]:
# Modifying the order_id values
orders_df.loc[10479,'ORDER_ID'] = '2232640.7A'
line_items_df.loc[51662,'ORDER_ID'] = '2232640.7A'
line_items_df.loc[51663,'ORDER_ID'] = '2232640.7A'

# Verifying that this order_id is no longer duplicated in the orders_df
orders_df[orders_df['ORDER_ID'] == duplicated_order_ids[3]]

MERCHANT_ID   ORDER_ID  SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
10478     7572140  2232640.7  2232640                 US                IN   

                 ORDER_DT FULFILLED_DT  REPRINT_FLAG SALES_CHANNEL_TYPE_ID  \
10478 2020-12-07 18:36:13          NaT           NaN                     2   

       TOTAL_COST  TOTAL_SHIPPING MERCHANT_REGISTERED_DT  SUB_IS_ACTIVE_FLAG  \
10478           0               0    2020-11-02 15:41:56                True   

                 SUB_PLAN SHIPMENT_CARRIER SHIPMENT_DELIVERD_DT  
10478  business_account_3              NaN                  NaT

In [45]:
# Verifying that this order_id is no longer duplicated in the line_items_df
line_items_df[line_items_df['ORDER_ID'] == duplicated_order_ids[3]]

ORDER_ID PRINT_PROVIDER_ID  PRODUCT_BRAND PRODUCT_TYPE ITEM_STATUS  \
8895  2232640.7                29         Gildan       Hoodie    canceled   
8896  2232640.7                51  Generic brand  Accessories    canceled   
8897  2232640.7                51  Generic brand  Accessories    canceled   

      QUANTITY  
8895         1  
8896         1  
8897         2

### ORDER_ID 1996849.47

In [46]:
# Checking the content of the rows
orders_df[orders_df['ORDER_ID'] == duplicated_order_ids[4]]

MERCHANT_ID    ORDER_ID  SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
8099     7368224  1996849.47  1996849                 US                NC   
9320     7368224  1996849.47  1996849                 US                MD   

                ORDER_DT        FULFILLED_DT  REPRINT_FLAG  \
8099 2020-12-27 12:42:36 2020-12-29 22:49:17           NaN   
9320 2020-11-13 21:20:55 2020-11-15 15:34:18           NaN   

     SALES_CHANNEL_TYPE_ID  TOTAL_COST  TOTAL_SHIPPING MERCHANT_REGISTERED_DT  \
8099                     4         491             424    2020-09-03 16:42:53   
9320                     4        1070             524    2020-09-03 16:42:53   

      SUB_IS_ACTIVE_FLAG            SUB_PLAN SHIPMENT_CARRIER  \
8099                True  business_account_3             USPS   
9320                True  business_account_3             USPS   

     SHIPMENT_DELIVERD_DT  
8099  2021-01-08 18:39:25  
9320  2020-11-18 21:06:22

The ORDER_DT, TOTAL_COST and TOTAL_SHIPPING values don't match at all. This is not a duplicated record, these are **different ORDER_IDs**.

In [47]:
# Checking the items that belong to this order
line_items_df[line_items_df['ORDER_ID'] == duplicated_order_ids[4]]

ORDER_ID PRINT_PROVIDER_ID  PRODUCT_BRAND PRODUCT_TYPE  \
12698  1996849.47                10  Generic brand   Home Decor   
37316  1996849.47                10  Generic brand  Accessories   

              ITEM_STATUS  QUANTITY  
12698  shipment_delivered         1  
37316  shipment_delivered         1

I'll modify the second order_id the same way I previously did.

In [48]:
# Modifying the order_id values
orders_df.loc[9320,'ORDER_ID'] = '1996849.47A'
line_items_df.loc[37316,'ORDER_ID'] = '1996849.47A'

# Verifying that this order_id is no longer duplicated in the orders_df
orders_df[orders_df['ORDER_ID'] == duplicated_order_ids[4]]

MERCHANT_ID    ORDER_ID  SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
8099     7368224  1996849.47  1996849                 US                NC   

                ORDER_DT        FULFILLED_DT  REPRINT_FLAG  \
8099 2020-12-27 12:42:36 2020-12-29 22:49:17           NaN   

     SALES_CHANNEL_TYPE_ID  TOTAL_COST  TOTAL_SHIPPING MERCHANT_REGISTERED_DT  \
8099                     4         491             424    2020-09-03 16:42:53   

      SUB_IS_ACTIVE_FLAG            SUB_PLAN SHIPMENT_CARRIER  \
8099                True  business_account_3             USPS   

     SHIPMENT_DELIVERD_DT  
8099  2021-01-08 18:39:25

In [49]:
# Verifying that this order_id is no longer duplicated in the line_items_df
line_items_df[line_items_df['ORDER_ID'] == duplicated_order_ids[4]]

ORDER_ID PRINT_PROVIDER_ID  PRODUCT_BRAND PRODUCT_TYPE  \
12698  1996849.47                10  Generic brand   Home Decor   

              ITEM_STATUS  QUANTITY  
12698  shipment_delivered         1

### ORDER_ID 2229889.102

In [50]:
# Checking the content of the rows
orders_df[orders_df['ORDER_ID'] == duplicated_order_ids[5]]

MERCHANT_ID     ORDER_ID  SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
4722     7559003  2229889.102  2229889                 FR               NaN   
5851     7559003  2229889.102  2229889                 FR               NaN   

                ORDER_DT        FULFILLED_DT  REPRINT_FLAG  \
4722 2020-12-03 02:27:39 2020-12-09 17:15:58           NaN   
5851 2020-11-12 01:43:43 2020-11-21 08:40:59           NaN   

     SALES_CHANNEL_TYPE_ID  TOTAL_COST  TOTAL_SHIPPING MERCHANT_REGISTERED_DT  \
4722                     6         333            1000    2020-10-28 09:58:01   
5851                     6         478             800    2020-10-28 09:58:01   

      SUB_IS_ACTIVE_FLAG            SUB_PLAN SHIPMENT_CARRIER  \
4722                True  business_account_3              NaN   
5851                True  business_account_3          unknown   

     SHIPMENT_DELIVERD_DT  
4722                  NaT  
5851                  NaT

The ORDER_DT, TOTAL_COST and TOTAL_SHIPPING values don't match at all. This is not a duplicated record, these are **different ORDER_IDs**.

In [51]:
# Checking the items that belong to this order
line_items_df[line_items_df['ORDER_ID'] == duplicated_order_ids[5]]

ORDER_ID PRINT_PROVIDER_ID  PRODUCT_BRAND PRODUCT_TYPE ITEM_STATUS  \
34550  2229889.102                 9  Generic brand          Mug   fulfilled   
44327  2229889.102                48  Generic brand          Mug   fulfilled   

       QUANTITY  
34550         1  
44327         1

I'll modify the second order_id the same way I previously did.

In [52]:
# Modifying the order_id values
orders_df.loc[5851,'ORDER_ID'] = '2229889.102A'
line_items_df.loc[44327,'ORDER_ID'] = '2229889.102A'

# Verifying that this order_id is no longer duplicated in the orders_df
orders_df[orders_df['ORDER_ID'] == duplicated_order_ids[5]]

MERCHANT_ID     ORDER_ID  SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
4722     7559003  2229889.102  2229889                 FR               NaN   

                ORDER_DT        FULFILLED_DT  REPRINT_FLAG  \
4722 2020-12-03 02:27:39 2020-12-09 17:15:58           NaN   

     SALES_CHANNEL_TYPE_ID  TOTAL_COST  TOTAL_SHIPPING MERCHANT_REGISTERED_DT  \
4722                     6         333            1000    2020-10-28 09:58:01   

      SUB_IS_ACTIVE_FLAG            SUB_PLAN SHIPMENT_CARRIER  \
4722                True  business_account_3              NaN   

     SHIPMENT_DELIVERD_DT  
4722                  NaT

In [53]:
# Verifying that this order_id is no longer duplicated in the line_items_df
line_items_df[line_items_df['ORDER_ID'] == duplicated_order_ids[5]]

ORDER_ID PRINT_PROVIDER_ID  PRODUCT_BRAND PRODUCT_TYPE ITEM_STATUS  \
34550  2229889.102                 9  Generic brand          Mug   fulfilled   

       QUANTITY  
34550         1

### ORDER_ID 2078441.1

In [54]:
# Checking the content of the rows
orders_df[orders_df['ORDER_ID'] == duplicated_order_ids[6]]

MERCHANT_ID   ORDER_ID  SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
1534     7439351  2078441.1  2078441                 US                IA   
1535     7439351  2078441.1  2078441                 US                IA   

                ORDER_DT        FULFILLED_DT  REPRINT_FLAG  \
1534 2020-10-03 21:37:34 2020-10-20 08:02:12           NaN   
1535 2020-10-03 21:37:34 2020-10-20 08:02:12           NaN   

     SALES_CHANNEL_TYPE_ID  TOTAL_COST  TOTAL_SHIPPING MERCHANT_REGISTERED_DT  \
1534                     2        3664            1700    2020-09-23 23:42:43   
1535                     2        3664            1700    2020-09-23 23:42:43   

      SUB_IS_ACTIVE_FLAG SUB_PLAN SHIPMENT_CARRIER SHIPMENT_DELIVERD_DT  
1534               False      NaN             USPS  2020-10-08 19:47:38  
1535               False      NaN             USPS  2020-10-20 08:11:09

The ORDER_DT, TOTAL_COST and TOTAL_SHIPPING values match, this is a **duplicated record** because only the SHIPMENT_DELIVERD_DT is different. So there's no need to modify the ORDER_ID value in any row.

In [55]:
# Checking the items that belong to this order
line_items_df[line_items_df['ORDER_ID'] == duplicated_order_ids[6]]

ORDER_ID PRINT_PROVIDER_ID  PRODUCT_BRAND PRODUCT_TYPE  \
38647  2078441.1                10  Generic brand   Home Decor   
38648  2078441.1                10  Generic brand   Home Decor   
38649  2078441.1                 1  Generic brand          Mug   
38650  2078441.1                 1  Generic brand          Mug   

              ITEM_STATUS  QUANTITY  
38647  shipment_delivered         1  
38648  shipment_delivered         1  
38649  shipment_delivered         1  
38650  shipment_delivered         1

### ORDER_ID 2015444.3

In [56]:
# Checking the content of the rows
orders_df[orders_df['ORDER_ID'] == duplicated_order_ids[7]]

MERCHANT_ID   ORDER_ID  SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
10989     7383862  2015444.3  2015444                 US                PA   
10990     7383862  2015444.3  2015444                 US                PA   

                 ORDER_DT        FULFILLED_DT  REPRINT_FLAG  \
10989 2020-12-24 13:53:24 2021-01-21 23:56:17           NaN   
10990 2020-12-24 13:53:24 2021-01-21 23:56:17           NaN   

      SALES_CHANNEL_TYPE_ID  TOTAL_COST  TOTAL_SHIPPING  \
10989                     2       28172            7480   
10990                     2       28172            7480   

      MERCHANT_REGISTERED_DT  SUB_IS_ACTIVE_FLAG            SUB_PLAN  \
10989    2020-09-09 21:25:29                True  business_account_3   
10990    2020-09-09 21:25:29                True  business_account_3   

      SHIPMENT_CARRIER SHIPMENT_DELIVERD_DT  
10989             USPS  2021-01-11 16:36:38  
10990             USPS                  NaT

The ORDER_DT, TOTAL_COST and TOTAL_SHIPPING values match, this is a **duplicated record** because only the SHIPMENT_DELIVERD_DT is different. So there's no need to modify the ORDER_ID value in any row.

In [57]:
# Checking the items that belong to this order
line_items_df[line_items_df['ORDER_ID'] == duplicated_order_ids[7]]

ORDER_ID PRINT_PROVIDER_ID  PRODUCT_BRAND     PRODUCT_TYPE  \
35240  2015444.3                14  Generic brand  All Over Prints   
35241  2015444.3                14  Generic brand  All Over Prints   
35242  2015444.3                14  Generic brand  All Over Prints   
35243  2015444.3                14  Generic brand  All Over Prints   
35244  2015444.3                10  Generic brand  All Over Prints   
35245  2015444.3                10  Generic brand  All Over Prints   
35246  2015444.3                10  Generic brand  All Over Prints   
35247  2015444.3                10  Generic brand  All Over Prints   
35248  2015444.3                14  Generic brand  All Over Prints   
35249  2015444.3                14  Generic brand  All Over Prints   
35250  2015444.3                45     Next Level          T-Shirt   
35251  2015444.3                45     Next Level          T-Shirt   
35252  2015444.3                16  Generic brand          T-Shirt   
35253  2015444.3                16  Generic brand          T-Shirt   
35254  2015444.3                45     Next Level          T-Shirt   
35255  2015444.3                45     Next Level          T-Shirt   
35256  2015444.3                10  Generic brand             Bags   
35257  2015444.3                10  Generic brand             Bags   
35258  2015444.3                10  Generic brand             Bags   
35259  2015444.3                10  Generic brand             Bags   
35260  2015444.3                45     Next Level          T-Shirt   
35261  2015444.3                45     Next Level          T-Shirt   
35262  2015444.3                29     Next Level          T-Shirt   
35263  2015444.3                29     Next Level          T-Shirt   
35264  2015444.3                42     Next Level          T-Shirt   
35265  2015444.3                42     Next Level          T-Shirt   
35266  2015444.3                42     Next Level          T-Shirt   
35267  2015444.3                42     Next Level          T-Shirt   
35268  2015444.3                42     Next Level          T-Shirt   
35269  2015444.3                42     Next Level          T-Shirt   
35270  2015444.3                42     Next Level          T-Shirt   
35271  2015444.3                42     Next Level          T-Shirt   
35272  2015444.3                42     Next Level          T-Shirt   
35273  2015444.3                42     Next Level          T-Shirt   
35274  2015444.3                42     Next Level          T-Shirt   
35275  2015444.3                42     Next Level          T-Shirt   
35276  2015444.3                42     Next Level          T-Shirt   
35277  2015444.3                42     Next Level          T-Shirt   
35278  2015444.3                42     Next Level          T-Shirt   
35279  2015444.3                42     Next Level          T-Shirt   
35280  2015444.3                42     Next Level          T-Shirt   
35281  2015444.3                42     Next Level          T-Shirt   
35282  2015444.3                42     Next Level          T-Shirt   
35283  2015444.3                42     Next Level          T-Shirt   
35284  2015444.3                42     Next Level          T-Shirt   
35285  2015444.3                42     Next Level          T-Shirt   
35286  2015444.3                42     Next Level          T-Shirt   
35287  2015444.3                42     Next Level          T-Shirt   
35288  2015444.3                42     Next Level          T-Shirt   
35289  2015444.3                42     Next Level          T-Shirt   

              ITEM_STATUS  QUANTITY  
35240           fulfilled         1  
35241           fulfilled         1  
35242           fulfilled         1  
35243           fulfilled         1  
35244    shipment_failure         1  
35245    shipment_failure         1  
35246    shipment_failure         1  
35247    shipment_failure         1  
35248           fulfilled         1  
35249           fulfilled         1  
35250  shipment_del

I can see some items weren't delivered yet or had a failure. That would explain the difference in the SHIPMENT_DELIVERD_DT column of the orders_df.

### ORDER_ID 1820693.162

In [58]:
# Checking the content of the rows
orders_df[orders_df['ORDER_ID'] == duplicated_order_ids[8]]

MERCHANT_ID     ORDER_ID  SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
6069     7223323  1820693.162  1820693                 US                SC   
6070     7223323  1820693.162  1820693                 US                SC   

                ORDER_DT        FULFILLED_DT  REPRINT_FLAG  \
6069 2020-11-04 04:20:21 2020-11-09 18:34:04           NaN   
6070 2020-11-04 04:20:21 2020-11-09 18:34:04           NaN   

     SALES_CHANNEL_TYPE_ID  TOTAL_COST  TOTAL_SHIPPING MERCHANT_REGISTERED_DT  \
6069                     3        6855            2624    2020-07-06 19:27:58   
6070                     3        6855            2624    2020-07-06 19:27:58   

      SUB_IS_ACTIVE_FLAG            SUB_PLAN SHIPMENT_CARRIER  \
6069                True  business_account_3             USPS   
6070                True  business_account_3             USPS   

     SHIPMENT_DELIVERD_DT  
6069  2020-11-09 22:02:28  
6070  2020-11-13 20:28:54

The ORDER_DT, TOTAL_COST and TOTAL_SHIPPING values match, this is a **duplicated record** because only the SHIPMENT_DELIVERD_DT is different. So there's no need to modify the ORDER_ID value in any row.

In [59]:
# Checking the items that belong to this order
line_items_df[line_items_df['ORDER_ID'] == duplicated_order_ids[8]]

ORDER_ID PRINT_PROVIDER_ID PRODUCT_BRAND  PRODUCT_TYPE  \
5697  1820693.162                29  Bella+Canvas       T-Shirt   
5698  1820693.162                29  Bella+Canvas       T-Shirt   
5699  1820693.162                16        Gildan  Kids clothes   
5700  1820693.162                16        Gildan  Kids clothes   
5701  1820693.162                16        Gildan  Kids clothes   
5702  1820693.162                16        Gildan  Kids clothes   
5703  1820693.162                 3        Gildan       T-Shirt   
5704  1820693.162                 3        Gildan       T-Shirt   
5705  1820693.162                 3        Gildan       T-Shirt   
5706  1820693.162                 3        Gildan       T-Shirt   
5707  1820693.162                45        Gildan  Kids clothes   
5708  1820693.162                45        Gildan  Kids clothes   
5709  1820693.162                45        Gildan  Kids clothes   
5710  1820693.162                45        Gildan  Kids clothes   
5711  1820693.162                45        Gildan  Kids clothes   
5712  1820693.162                45        Gildan  Kids clothes   

             ITEM_STATUS  QUANTITY  
5697  shipment_delivered         1  
5698  shipment_delivered         1  
5699           fulfilled         1  
5700           fulfilled         1  
5701           fulfilled         1  
5702           fulfilled         1  
5703  shipment_delivered         1  
5704  shipment_delivered         1  
5705  shipment_delivered         1  
5706  shipment_delivered         1  
5707  shipment_delivered         1  
5708  shipment_delivered         1  
5709  shipment_delivered         1  
5710  shipment_delivered         1  
5711  shipment_delivered         1  
5712  shipment_delivered         1

### ORDER_ID 363596.227	

In [60]:
# Checking the content of the rows
orders_df[orders_df['ORDER_ID'] == duplicated_order_ids[9]]

MERCHANT_ID    ORDER_ID SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
537     5868230  363596.227  363596                 US                VI   
538     5868230  363596.227  363596                 US                VI   

               ORDER_DT        FULFILLED_DT  REPRINT_FLAG  \
537 2020-11-18 19:36:02 2020-11-19 15:23:46           NaN   
538 2020-11-18 19:36:02 2020-11-19 15:23:46           NaN   

    SALES_CHANNEL_TYPE_ID  TOTAL_COST  TOTAL_SHIPPING MERCHANT_REGISTERED_DT  \
537                     2        9533            2616    2018-01-16 11:29:38   
538                     2        9533            2616    2018-01-16 11:29:38   

     SUB_IS_ACTIVE_FLAG            SUB_PLAN SHIPMENT_CARRIER  \
537                True  business_account_3             USPS   
538                True  business_account_3             USPS   

    SHIPMENT_DELIVERD_DT  
537  2020-11-24 17:34:42  
538  2020-11-28 16:51:06

The ORDER_DT, TOTAL_COST and TOTAL_SHIPPING values match, this is a **duplicated record** because only the SHIPMENT_DELIVERD_DT is different. So there's no need to modify the ORDER_ID value in any row.

In [61]:
# Checking the items that belong to this order
line_items_df[line_items_df['ORDER_ID'] == duplicated_order_ids[9]]

ORDER_ID PRINT_PROVIDER_ID  PRODUCT_BRAND    PRODUCT_TYPE  \
13615  363596.227                 1  Generic brand  Paper products   
13616  363596.227                 1  Generic brand  Paper products   
13617  363596.227                 1  Generic brand  Paper products   
13618  363596.227                 1  Generic brand  Paper products   
13619  363596.227                29   Bella+Canvas         T-Shirt   
13620  363596.227                29   Bella+Canvas         T-Shirt   
13621  363596.227                29   Bella+Canvas         T-Shirt   
13622  363596.227                29   Bella+Canvas         T-Shirt   
13623  363596.227                29   Bella+Canvas         T-Shirt   
13624  363596.227                29   Bella+Canvas         T-Shirt   
13625  363596.227                29   Bella+Canvas         T-Shirt   
13626  363596.227                29   Bella+Canvas         T-Shirt   

              ITEM_STATUS  QUANTITY  
13615  shipment_delivered         4  
13616  shipment_delivered         4  
13617  shipment_delivered         3  
13618  shipment_delivered         3  
13619  shipment_delivered         2  
13620  shipment_delivered         2  
13621  shipment_delivered         3  
13622  shipment_delivered         3  
13623  shipment_delivered         3  
13624  shipment_delivered         3  
13625  shipment_delivered         2  
13626  shipment_delivered         2

### ORDER_ID 1555863.282	

In [62]:
# Checking the content of the rows
orders_df[orders_df['ORDER_ID'] == duplicated_order_ids[10]]

MERCHANT_ID     ORDER_ID  SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
5285     7016547  1555863.282  1555863                 US                LA   
7502     7016547  1555863.282  1555863                 US                TX   

                ORDER_DT        FULFILLED_DT  REPRINT_FLAG  \
5285 2020-10-11 01:33:06 2020-10-14 21:42:00           NaN   
7502 2020-10-29 22:10:53 2020-11-02 18:07:59           NaN   

     SALES_CHANNEL_TYPE_ID  TOTAL_COST  TOTAL_SHIPPING MERCHANT_REGISTERED_DT  \
5285                     1         103             350    2020-02-17 23:36:26   
7502                     1         167             350    2020-02-17 23:36:26   

      SUB_IS_ACTIVE_FLAG            SUB_PLAN SHIPMENT_CARRIER  \
5285                True  business_account_3              NaN   
7502                True  business_account_3             USPS   

     SHIPMENT_DELIVERD_DT  
5285                  NaT  
7502  2020-11-05 20:05:27

The ORDER_DT and TOTAL_COST values don't match at all. This is not a duplicated record, these are **different ORDER_IDs**.

In [63]:
# Checking the items that belong to this order
line_items_df[line_items_df['ORDER_ID'] == duplicated_order_ids[10]]

ORDER_ID PRINT_PROVIDER_ID  PRODUCT_BRAND    PRODUCT_TYPE  \
54634  1555863.282                 1  Generic brand  Paper products   
57372  1555863.282                 1  Generic brand  Paper products   

              ITEM_STATUS  QUANTITY  
54634           fulfilled         1  
57372  shipment_delivered         1

I'll modify the second order_id the same way I previously did.

In [64]:
# Modifying the order_id values
orders_df.loc[7502,'ORDER_ID'] = '1555863.282A'
line_items_df.loc[57372,'ORDER_ID'] = '1555863.282A'

# Verifying that this order_id is no longer duplicated in the orders_df
orders_df[orders_df['ORDER_ID'] == duplicated_order_ids[10]]

MERCHANT_ID     ORDER_ID  SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
5285     7016547  1555863.282  1555863                 US                LA   

                ORDER_DT        FULFILLED_DT  REPRINT_FLAG  \
5285 2020-10-11 01:33:06 2020-10-14 21:42:00           NaN   

     SALES_CHANNEL_TYPE_ID  TOTAL_COST  TOTAL_SHIPPING MERCHANT_REGISTERED_DT  \
5285                     1         103             350    2020-02-17 23:36:26   

      SUB_IS_ACTIVE_FLAG            SUB_PLAN SHIPMENT_CARRIER  \
5285                True  business_account_3              NaN   

     SHIPMENT_DELIVERD_DT  
5285                  NaT

In [65]:
# Checking the items that belong to this order
line_items_df[line_items_df['ORDER_ID'] == duplicated_order_ids[10]]

ORDER_ID PRINT_PROVIDER_ID  PRODUCT_BRAND    PRODUCT_TYPE  \
54634  1555863.282                 1  Generic brand  Paper products   

      ITEM_STATUS  QUANTITY  
54634   fulfilled         1

Now that I finished verifying the duplicated records and modifying the values of the records that were different, I want to **get the order_ids that are indeed duplicated**.

In [66]:
# Getting the order_ids that are duplicated
orders_df['ORDER_ID'].value_counts()[orders_df['ORDER_ID'].value_counts() > 1]

516066.48428    2
363596.227      2
752154.3        2
2015444.3       2
1820693.162     2
2078441.1       2
Name: ORDER_ID, dtype: int64

As I previously mentioned, a solution to avoid inaccurate results when doing aggregations on this df would be to create a new column indicating that these records have a duplicate.

### **Adding a new column to orders_df**

In [67]:
# I'll first add a new column with the constant value of 'No' for each row
orders_df['DUPLICATED_ORDER'] = 'No'
orders_df.head()

MERCHANT_ID     ORDER_ID  SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
0     7143578  2013244.178  2013244                 US                WA   
1     5888534  1358429.695  1358429                 US                MN   
2     6484224   939390.193   939390                 US                AK   
3     7251033    1856107.4  1856107                 US                PA   
4     7241488    1843904.4  1843904                 US                CO   

             ORDER_DT        FULFILLED_DT  REPRINT_FLAG SALES_CHANNEL_TYPE_ID  \
0 2020-10-02 19:30:15 2020-10-20 07:53:06           NaN                     9   
1 2020-11-21 05:19:03 2020-11-23 00:23:32           NaN                     4   
2 2020-10-25 01:39:08 2020-10-26 20:30:53           NaN                     1   
3 2020-10-08 16:35:21 2020-10-10 15:02:47           NaN                     1   
4 2020-11-17 19:30:10 2020-11-18 03:45:25           NaN                     2   

   TOTAL_COST  TOTAL_SHIPPING MERCHANT_REGISTERED_DT  SUB_IS_ACTIVE_FLAG  \
0        2902            2200    2020-05-19 15:22:57                True   
1        1149             924    2018-02-08 07:34:02                True   
2         977             400    2019-04-09 20:02:46                True   
3         661             800    2020-07-20 17:04:41                True   
4         719             424    2020-07-15 18:47:51                True   

             SUB_PLAN SHIPMENT_CARRIER SHIPMENT_DELIVERD_DT DUPLICATED_ORDER  
0  business_account_3              UPS  2020-10-08 21:07:29               No  
1  business_account_3             USPS  2020-11-28 21:50:56               No  
2  business_account_3             USPS  2020-11-02 23:37:54               No  
3  business_account_3              UPS  2020-10-15 18:08:37               No  
4  business_account_3             USPS  2020-11-24 00:44:56               No

In [68]:
# Next, I want to check the rows that have duplicated ORDER_IDs in the table
duplicated_order_ids = list(orders_df['ORDER_ID'].value_counts()[orders_df['ORDER_ID'].value_counts() > 1].index)
orders_df[orders_df['ORDER_ID'].isin(duplicated_order_ids)]

MERCHANT_ID      ORDER_ID  SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
13        6314037      752154.3   752154                 US                WA   
14        6314037      752154.3   752154                 US                WA   
537       5868230    363596.227   363596                 US                VI   
538       5868230    363596.227   363596                 US                VI   
1534      7439351     2078441.1  2078441                 US                IA   
1535      7439351     2078441.1  2078441                 US                IA   
5436      6118692  516066.48428   516066                 US                MD   
5437      6118692  516066.48428   516066                 US                MD   
6069      7223323   1820693.162  1820693                 US                SC   
6070      7223323   1820693.162  1820693                 US                SC   
10989     7383862     2015444.3  2015444                 US                PA   
10990     7383862     2015444.3  2015444                 US                PA   

                 ORDER_DT        FULFILLED_DT  REPRINT_FLAG  \
13    2020-10-03 13:05:53 2020-10-07 20:20:59           NaN   
14    2020-10-03 13:05:53 2020-10-07 20:20:59           NaN   
537   2020-11-18 19:36:02 2020-11-19 15:23:46           NaN   
538   2020-11-18 19:36:02 2020-11-19 15:23:46           NaN   
1534  2020-10-03 21:37:34 2020-10-20 08:02:12           NaN   
1535  2020-10-03 21:37:34 2020-10-20 08:02:12           NaN   
5436  2020-10-11 04:30:55 2020-10-13 01:11:15           NaN   
5437  2020-10-11 04:30:55 2020-10-13 01:11:15           NaN   
6069  2020-11-04 04:20:21 2020-11-09 18:34:04           NaN   
6070  2020-11-04 04:20:21 2020-11-09 18:34:04           NaN   
10989 2020-12-24 13:53:24 2021-01-21 23:56:17           NaN   
10990 2020-12-24 13:53:24 2021-01-21 23:56:17           NaN   

      SALES_CHANNEL_TYPE_ID  TOTAL_COST  TOTAL_SHIPPING  \
13                        2       14732            3884   
14                        2       14732            3884   
537                       2        9533            2616   
538                       2        9533            2616   
1534                      2        3664            1700   
1535                      2        3664            1700   
5436                      1        1979             850   
5437                      1        1979             850   
6069                      3        6855            2624   
6070                      3        6855            2624   
10989                     2       28172            7480   
10990                     2       28172            7480   

      MERCHANT_REGISTERED_DT  SUB_IS_ACTIVE_FLAG            SUB_PLAN  \
13       2019-01-25 17:38:56               False  business_account_3   
14       2019-01-25 17:38:56               False  business_account_3   
537      2018-01-16 11:29:38                True  business_account_3   
538      2018-01-16 11:29:38                True  business_account_3   
1534     2020-09-23 23:42:43               False                 NaN   
1535     2020-09-23 23:42:43               False                 NaN   
5436     2018-09-13 12:23:50                True  business_account_3   
5437     2018-09-13 12:23:50                True  business_account_3   
6069     2020-07-06 19:27:58                True  business_account_3   
6070     2020-07-06 19:27:58                True  business_account_3   
10989    2020-09-09 21:25:29                True  business_account_3   
10990    2020-09-09 21:25:29                True  business_account_3   

           SHIPMENT_CARRIER SHIPMENT_DELIVERD_DT DUPLICATED_ORDER  
13                     USPS  2020-10-09 21:38:49               No  
14                      NaN                  NaT               No  
537                    USPS  2020-11-24 17:34:42               No  
538                    USPS  2020-11-28 16:51:06               No  
1534                   USPS  2020-10-08 19:47:38               No  
1535               

In [69]:
# Now I'll create a list that will contain only the indexes of the duplicated records that are displayed in the cell above
index_duplic = []
index_number = 0
for i in list(orders_df[orders_df['ORDER_ID'].isin(duplicated_order_ids)].index):
    if index_number%2 != 0:
        index_duplic.append(i)
    index_number = index_number + 1

index_duplic

[14, 538, 1535, 5437, 6070, 10990]

In [70]:
# Finally, I will change the value of the recently created column from 'No' to 'Yes' for the indexes I just got
for i in index_duplic:
    orders_df.loc[i,'DUPLICATED_ORDER'] = 'Yes'

# Verifying if the changed indeed happened
orders_df[orders_df['ORDER_ID'].isin(duplicated_order_ids)]

MERCHANT_ID      ORDER_ID  SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
13        6314037      752154.3   752154                 US                WA   
14        6314037      752154.3   752154                 US                WA   
537       5868230    363596.227   363596                 US                VI   
538       5868230    363596.227   363596                 US                VI   
1534      7439351     2078441.1  2078441                 US                IA   
1535      7439351     2078441.1  2078441                 US                IA   
5436      6118692  516066.48428   516066                 US                MD   
5437      6118692  516066.48428   516066                 US                MD   
6069      7223323   1820693.162  1820693                 US                SC   
6070      7223323   1820693.162  1820693                 US                SC   
10989     7383862     2015444.3  2015444                 US                PA   
10990     7383862     2015444.3  2015444                 US                PA   

                 ORDER_DT        FULFILLED_DT  REPRINT_FLAG  \
13    2020-10-03 13:05:53 2020-10-07 20:20:59           NaN   
14    2020-10-03 13:05:53 2020-10-07 20:20:59           NaN   
537   2020-11-18 19:36:02 2020-11-19 15:23:46           NaN   
538   2020-11-18 19:36:02 2020-11-19 15:23:46           NaN   
1534  2020-10-03 21:37:34 2020-10-20 08:02:12           NaN   
1535  2020-10-03 21:37:34 2020-10-20 08:02:12           NaN   
5436  2020-10-11 04:30:55 2020-10-13 01:11:15           NaN   
5437  2020-10-11 04:30:55 2020-10-13 01:11:15           NaN   
6069  2020-11-04 04:20:21 2020-11-09 18:34:04           NaN   
6070  2020-11-04 04:20:21 2020-11-09 18:34:04           NaN   
10989 2020-12-24 13:53:24 2021-01-21 23:56:17           NaN   
10990 2020-12-24 13:53:24 2021-01-21 23:56:17           NaN   

      SALES_CHANNEL_TYPE_ID  TOTAL_COST  TOTAL_SHIPPING  \
13                        2       14732            3884   
14                        2       14732            3884   
537                       2        9533            2616   
538                       2        9533            2616   
1534                      2        3664            1700   
1535                      2        3664            1700   
5436                      1        1979             850   
5437                      1        1979             850   
6069                      3        6855            2624   
6070                      3        6855            2624   
10989                     2       28172            7480   
10990                     2       28172            7480   

      MERCHANT_REGISTERED_DT  SUB_IS_ACTIVE_FLAG            SUB_PLAN  \
13       2019-01-25 17:38:56               False  business_account_3   
14       2019-01-25 17:38:56               False  business_account_3   
537      2018-01-16 11:29:38                True  business_account_3   
538      2018-01-16 11:29:38                True  business_account_3   
1534     2020-09-23 23:42:43               False                 NaN   
1535     2020-09-23 23:42:43               False                 NaN   
5436     2018-09-13 12:23:50                True  business_account_3   
5437     2018-09-13 12:23:50                True  business_account_3   
6069     2020-07-06 19:27:58                True  business_account_3   
6070     2020-07-06 19:27:58                True  business_account_3   
10989    2020-09-09 21:25:29                True  business_account_3   
10990    2020-09-09 21:25:29                True  business_account_3   

           SHIPMENT_CARRIER SHIPMENT_DELIVERD_DT DUPLICATED_ORDER  
13                     USPS  2020-10-09 21:38:49               No  
14                      NaN                  NaT              Yes  
537                    USPS  2020-11-24 17:34:42               No  
538                    USPS  2020-11-28 16:51:06              Yes  
1534                   USPS  2020-10-08 19:47:38               No  
1535               

With this addition to the orders_df I can now move on to the rest of the columns.

## Working on the SHOP_ID column

In [71]:
# Looking for Nulls in the SHOP_ID column
sum(orders_df['SHOP_ID'].isna())

0

In [72]:
# Getting a glance of the data in this column
orders_df['SHOP_ID'].value_counts()

1559676    150
105752      81
1280396     81
2151136     73
1750015     71
          ... 
1788929      1
1771787      1
1455007      1
2080651      1
1717607      1
Name: SHOP_ID, Length: 5049, dtype: int64

In [73]:
# Based on the previous result, I'd like to know to number of unique values for this column
orders_df['SHOP_ID'].nunique()

5049

## Working on the ADDRESS_TO_COUNTRY column

In [74]:
# Looking for Nulls in the ADDRESS_TO_COUNTRY column
sum(orders_df['ADDRESS_TO_COUNTRY'].isna())

1

In [75]:
# Locating the Null value
orders_df[orders_df['ADDRESS_TO_COUNTRY'].isna()]

MERCHANT_ID    ORDER_ID  SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
10430     6862652  1374085.21  1374085                NaN               NaN   

                 ORDER_DT FULFILLED_DT  REPRINT_FLAG SALES_CHANNEL_TYPE_ID  \
10430 2020-10-04 21:39:15          NaT           NaN                     3   

       TOTAL_COST  TOTAL_SHIPPING MERCHANT_REGISTERED_DT  SUB_IS_ACTIVE_FLAG  \
10430        1045               0    2019-10-13 19:54:26               False   

      SUB_PLAN SHIPMENT_CARRIER SHIPMENT_DELIVERD_DT DUPLICATED_ORDER  
10430      NaN              NaN                  NaT               No

This record has a lot of Nulls across the different fileds of the table. I should also check the line_items_df associated with this order_id to see what happened to the order.

In [76]:
# Locating this order_id in the line_items_df
line_items_df[line_items_df['ORDER_ID'] == orders_df[orders_df['ADDRESS_TO_COUNTRY'].isna()]['ORDER_ID'].iloc[0]]

ORDER_ID PRINT_PROVIDER_ID  PRODUCT_BRAND     PRODUCT_TYPE  \
29789  1374085.21                33  Generic brand  All Over Prints   

      ITEM_STATUS  QUANTITY  
29789    canceled         1

Looks like the reason for so many Nulls in this record is because the order was cancelled. However, NaNs are not properly read by SQL, so I'll filled the NaNs with a blank space(''), after doing so I'll move on with the cleaning. Later in SQL I'll change these blank spaces to proper Nulls. This is the approach I'll take for NaNs in String columns.

In [77]:
# Making the change
orders_df['ADDRESS_TO_COUNTRY'] = orders_df['ADDRESS_TO_COUNTRY'].fillna('')

# Verifying that the changed happened
sum(orders_df['ADDRESS_TO_COUNTRY'].isna())

0

In [78]:
# Getting a glance of the data in this column
orders_df['ADDRESS_TO_COUNTRY'].value_counts()

US    11816
CA      515
GB      416
DE      261
AU       89
      ...  
TN        1
GI        1
BY        1
KR        1
BL        1
Name: ADDRESS_TO_COUNTRY, Length: 67, dtype: int64

In [79]:
# Based on the previous result, I'd like to know to number of unique values for this column
orders_df['ADDRESS_TO_COUNTRY'].nunique()

67

## Working on the ADDRESS_TO_REGION column

In [80]:
# Looking for Nulls in the ADDRESS_TO_REGION column
sum(orders_df['ADDRESS_TO_REGION'].isna())

576

In [81]:
# Locating the Null value
orders_df[orders_df['ADDRESS_TO_REGION'].isna()].head()

MERCHANT_ID       ORDER_ID  SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
9       5900409    1325831.102  1325831                 BE               NaN   
32      7520217     2170880.11  2170880                 DK               NaN   
55      7271254     1881174.25  1881174                 DE               NaN   
97      7501649     2148728.46  2148728                 DE               NaN   
154     6932458  1449381.11828  1449381                 GB               NaN   

               ORDER_DT        FULFILLED_DT  REPRINT_FLAG  \
9   2020-11-11 11:19:05 2020-11-12 02:44:37           NaN   
32  2020-11-18 14:07:45 2020-11-24 14:39:25           NaN   
55  2020-11-20 14:43:59 2020-11-23 16:23:32           NaN   
97  2020-11-17 17:34:43 2020-11-20 02:36:14           NaN   
154 2020-12-05 19:10:06 2020-12-13 15:10:16           NaN   

    SALES_CHANNEL_TYPE_ID  TOTAL_COST  TOTAL_SHIPPING MERCHANT_REGISTERED_DT  \
9                       1        2880            1180    2018-02-23 06:30:55   
32                      1         889             400    2020-10-15 09:48:46   
55                      4        1761            2690    2020-07-29 20:01:36   
97                      2        1540            1200    2020-10-08 15:49:59   
154                     1        1024             800    2019-12-11 22:10:14   

     SUB_IS_ACTIVE_FLAG            SUB_PLAN      SHIPMENT_CARRIER  \
9                  True              Plan 4                   NaN   
32                False  business_account_3                   NaN   
55                False  business_account_3                   NaN   
97                False  business_account_3          GLOBEGISTICS   
154                True              Plan 4  UPS_MAIL_INNOVATIONS   

    SHIPMENT_DELIVERD_DT DUPLICATED_ORDER  
9                    NaT               No  
32                   NaT               No  
55                   NaT               No  
97                   NaT               No  
154  2021-01-22 15:13:23               No

In [82]:
# I want to know if these records share anything in common, I will start by checking the country they belong to
orders_df[orders_df['ADDRESS_TO_REGION'].isna()]['ADDRESS_TO_COUNTRY'].value_counts()

GB    203
DE    187
FR     52
BE     13
CH     13
SE     13
DK     12
AT     11
NL     11
CZ     11
FI      6
IL      5
NO      5
HU      5
SG      4
SA      3
LT      2
EE      2
PL      2
GR      2
SI      2
PR      1
IT      1
ES      1
SK      1
TN      1
LU      1
        1
MO      1
LV      1
MU      1
RS      1
CG      1
Name: ADDRESS_TO_COUNTRY, dtype: int64

In [83]:
# It looks like whenever I have a Null in the 'ADDRESS_TO_REGION' column the Country is not the US. I'll verify that
'US' in list(orders_df[orders_df['ADDRESS_TO_REGION'].isna()]['ADDRESS_TO_COUNTRY'].value_counts().index)

False

In [84]:
# Now I'd like to know if I can have values for the ADDRESS_TO_REGION column even when the country is not the US
orders_df[orders_df['ADDRESS_TO_REGION'].isna() == False]['ADDRESS_TO_COUNTRY'].value_counts().head()

US    11816
CA      515
GB      213
AU       89
DE       74
Name: ADDRESS_TO_COUNTRY, dtype: int64

In [85]:
orders_df[orders_df['ADDRESS_TO_COUNTRY'] == 'GB'].head()

MERCHANT_ID       ORDER_ID  SHOP_ID ADDRESS_TO_COUNTRY ADDRESS_TO_REGION  \
21      2844040      2254663.2  2254663                 GB     EDINBURGHCITY   
31      7453498    2094464.195  2094464                 GB               ENG   
101     6860545    1366702.279  1366702                 GB        SHROPSHIRE   
121     7019548  1559676.23836  1559676                 GB           ENGLAND   
130     6931939    1448783.133  1448783                 GB           ENGLAND   

               ORDER_DT        FULFILLED_DT  REPRINT_FLAG  \
21  2020-11-09 11:07:39 2020-11-12 20:25:26           NaN   
31  2020-11-05 15:05:47 2020-11-11 21:56:21           NaN   
101 2020-12-04 16:08:49 2020-12-11 22:08:00           NaN   
121 2020-11-26 08:24:16 2020-11-27 10:00:49           NaN   
130 2020-11-26 20:17:49 2020-11-27 17:33:12           NaN   

    SALES_CHANNEL_TYPE_ID  TOTAL_COST  TOTAL_SHIPPING MERCHANT_REGISTERED_DT  \
21                      4         333             560    2017-05-31 13:58:01   
31                      9         944             459    2020-09-27 14:57:40   
101                     4        1115             550    2019-10-11 15:57:25   
121                     6         572             940    2020-02-20 01:09:32   
130                     4        2331            1070    2019-12-11 14:37:40   

     SUB_IS_ACTIVE_FLAG            SUB_PLAN SHIPMENT_CARRIER  \
21                 True  business_account_3              NaN   
31                 True  business_account_3       ROYAL_MAIL   
101                True  business_account_3       ROYAL_MAIL   
121                True  business_account_3              NaN   
130               False                 NaN              NaN   

    SHIPMENT_DELIVERD_DT DUPLICATED_ORDER  
21                   NaT               No  
31                   NaT               No  
101                  NaT               No  
121                  NaT               No  
130                  NaT               No

The reason I have Nulls in the ADDRESS_TO_REGION is because the country of delivery is not the US. As seen in the previous cell having a different country of destination than the US can result in a Null for the ADDRESS_TO_REGION column, but it can also have a value. I'll change those nulls to blank spaces now.

In [86]:
# Making the change
orders_df['ADDRESS_TO_REGION'] = orders_df['ADDRESS_TO_REGION'].fillna('')

# Verifying that the changed happened
sum(orders_df['ADDRESS_TO_REGION'].isna())

0

In [87]:
# Getting a glance of the data in this column
orders_df['ADDRESS_TO_REGION'].value_counts()

CA             1576
TX              810
NY              761
FL              636
                576
               ... 
LIÈGE            1
NETHERLANDS       1
PR                1
PUERTO RICO       1
TOKYO             1
Name: ADDRESS_TO_REGION, Length: 395, dtype: int64

In [88]:
# Based on the previous result, I'd like to know to number of unique values for this column
orders_df['ADDRESS_TO_REGION'].nunique()

395

## Working on the ORDER_DT column

In [89]:
# Looking for Nulls in the ORDER_DT column
sum(orders_df['ORDER_DT'].isna())

0

In [90]:
# Getting a glance of the data in this column
orders_df['ORDER_DT'].value_counts()

2020-12-03 08:46:27    2
2020-12-14 11:57:48    2
2020-12-09 22:38:42    2
2020-11-21 23:08:21    2
2020-12-03 17:08:43    2
                      ..
2020-12-05 00:11:07    1
2020-12-02 20:45:48    1
2020-12-04 09:12:34    1
2020-12-17 01:33:38    1
2020-11-26 00:33:02    1
Name: ORDER_DT, Length: 13479, dtype: int64

In [91]:
# Locating the records that share the same ORDER_DT and sorting the records based on that value
orders_df[orders_df['ORDER_DT'].isin(
    list(orders_df['ORDER_DT'].value_counts()[orders_df['ORDER_DT'].value_counts() > 1].index))].sort_values('ORDER_DT')

MERCHANT_ID       ORDER_ID  SHOP_ID ADDRESS_TO_COUNTRY  \
13        6314037       752154.3   752154                 US   
14        6314037       752154.3   752154                 US   
1534      7439351      2078441.1  2078441                 US   
1535      7439351      2078441.1  2078441                 US   
8922      7328199   1951270.1136  1951270                 US   
8267      7101973    1663376.632  1663376                 CA   
10026     6898316   1409941.1943  1409941                 US   
4106      6922494    1437788.709  1437788                 US   
308       7200586    1790915.736  1790915                 US   
7510      7110501   1674237.9262  1674237                 US   
11379     7110501   1674237.9668  1674237                 US   
4592      5889884      253502.97   253502                 US   
5437      6118692   516066.48428   516066                 US   
5436      6118692   516066.48428   516066                 US   
6070      7223323    1820693.162  1820693                 US   
6069      7223323    1820693.162  1820693                 US   
2882      5789144   1269541.1997  1269541                 US   
5851      7559003   2229889.102A  2229889                 FR   
6551      7186353    1772049.255  1772049                 US   
3354      7184859    1770064.134  1770064                 US   
538       5868230     363596.227   363596                 US   
537       5868230     363596.227   363596                 US   
5272      7513718     2163120.46  2163120                 US   
3195      7208306   1800982.1394  1800982                 SE   
8966      6341646     782724.589   782724                 US   
13244     7185125   1840241.1755  1840241                 US   
12649     6906823   1419673.6221  1419673                 US   
824       7087244      1644501.5  1644501                 US   
7290      7244962    1848395.768  1848395                 US   
7103      6209058    624552.6608   624552                 US   
603       6246822  1225155.10575  1225155                 US   
3208      6795017  1280396.38169  1280396                 US   
4879      7154418     1730644.93  1730644                 US   
10798     7194519   1782844.6792  1782844                 US   
6159      7085857   1642717.1335  1642717                 US   
9764      7403652     2039092.13  2039092                 US   
8076      6433350   1637235.7416  1637235                 US   
10787     6433350    1637235.742  1637235                 US   
4607      7244522     1847818.72  1847818                 US   
10318     6622512     1088988.34  1088988                 US   
2801      7307615    1924745.717  1924745                 US   
8732      7156339   1733523.2189  1733523                 GB   
4382      6586671    1179258.197  1179258                 US   
4510      7166263    1745827.854  1745827                 US   
10270     5900409  1325831.11976  1325831                 IE   
1156      7142840    1719569.835  1719569                 CA   
9042      7512808      2162041.7  2162041                 US   
9602      7201001   1791385.3939  1791385                 US   
8629      7164911    1744378.264  1744378                 US   
7496      6839466    1337199.155  1337199                 US   
6826      6956453   1479700.1188  1479700                 US   
4442      7148745     1723364.59  1723364                 US   
10989     7383862      2015444.3  2015444                 US   
10990     7383862      2015444.3  2015444                 US   

      ADDRESS_TO_REGION            ORDER_DT        FULFILLED_DT  REPRINT_FLAG  \
13                   WA 2020-10-03 13:05:53 2020-10-07 20:20:59           NaN   
14                   WA 2020-10-03 13:05:53 2020-10-07 20:20:59           NaN   
1534                 IA 2020-10-03 21:37:34 2020-10-20 08:02:12           NaN   
1535                 IA 2020-10-03 21:37:34 2020-10-20 08:02:12           NaN   
8922                 GA 2020-10-05 02:01:09 2020-10-12 18:21:56           NaN   

After eyeballing these records I can tell that some of the records with the same ORDER_DT are the duplicated rows I previously worked on, and some others are just coincidences. There's not much to this column.

## Working on the FULFILLED_DT column

In [92]:
# Looking for Nulls in the FULFILLED_DT column
sum(orders_df['FULFILLED_DT'].isna())

869

Since this column refers to the Timestamp when the order was printed and some order status have not been printed yet, then it's ok to have Nulls. However, when I import this data to SQL Server it will generate a problem because the NaNs will be read as strigns and not as SQL Nulls so I'll replace the NaNs with 0, or its equivalent of '1900-01-01 00:00:01'. This is the approach I'll take with NaNs in datetime columns. In SQL Server I'll change these values to proper Nulls.

In [93]:
# Making the change
orders_df['FULFILLED_DT'] = orders_df['FULFILLED_DT'].fillna('1900-01-01 00:00:01')

# Verifying that the changed happened
sum(orders_df['FULFILLED_DT'].isna())

0

## Working on the REPRINT_FLAG column

In [94]:
# I want to see the REPRINT_FLAG column with more deatil
orders_df['REPRINT_FLAG'].value_counts()

1.0    103
Name: REPRINT_FLAG, dtype: int64

In [95]:
# I know I have over 13K rows, so if I only have 103 records with values, the rest must be blanks
sum(orders_df['REPRINT_FLAG'].isna())

13403

In [96]:
# REPRINT_FLAG is a boolean column, it has blanks but the records with values in the Excel file show 'True' as the value
# However, booleans are considered 1.0 or 0, So I'll convert this column to a string type with 'True' and 'False' as values
orders_df['REPRINT_FLAG'] = orders_df['REPRINT_FLAG'].map({1.0:'True'}).fillna('False')

In [97]:
# Verifying that the change indeed happened
orders_df['REPRINT_FLAG'].value_counts()

False    13403
True       103
Name: REPRINT_FLAG, dtype: int64

In [98]:
# Checking the datatypes again to see if the REPRINT_FLAG was casted as a string data type
orders_df['REPRINT_FLAG'].apply(type).value_counts()

<class 'str'>    13506
Name: REPRINT_FLAG, dtype: int64

## Working on the SALES_CHANNEL_TYPE_ID column

In [99]:
# Looking for Nulls in the SALES_CHANNEL_TYPE_ID column
sum(orders_df['SALES_CHANNEL_TYPE_ID'].isna())

0

In [100]:
# Getting a glance of the data in this column
orders_df['SALES_CHANNEL_TYPE_ID'].value_counts()

4     5372
2     4150
1     3278
6      298
3      203
9      172
7       26
11       7
Name: SALES_CHANNEL_TYPE_ID, dtype: int64

In [101]:
# Based on the previous result, I'd like to know to number of unique values for this column
orders_df['SALES_CHANNEL_TYPE_ID'].nunique()

8

## Working on the TOTAL_COST column

In [102]:
# Looking for Nulls in the TOTAL_COST column
sum(orders_df['TOTAL_COST'].isna())

0

Since this column refers to the costs, I don't really need to see the different values of this column, so I can move on to the next one.

## Working on the TOTAL_SHIPPING column

In [103]:
# Looking for Nulls in the TOTAL_SHIPPING column
sum(orders_df['TOTAL_SHIPPING'].isna())

0

Since this column refers to the costs, I don't really need to see the different values of this column, so I can move on to the next one.

## Working on the MERCHANT_REGISTERED_DT column

In [104]:
# Looking for Nulls in the MERCHANT_REGISTERED_DT column
sum(orders_df['MERCHANT_REGISTERED_DT'].isna())

0

Since this column refers to datetimes, I don't really need to see the different values of this column, so I can move on to the next one.

## Working on the SUB_IS_ACTIVE_FLAG column

In [105]:
# Looking for Nulls in the TOTAL_COST column
sum(orders_df['SUB_IS_ACTIVE_FLAG'].isna())

0

In [106]:
# Getting a glance of the data in this column
orders_df['SUB_IS_ACTIVE_FLAG'].value_counts()

True     9842
False    3664
Name: SUB_IS_ACTIVE_FLAG, dtype: int64

This is a boolean column, since it has no Nulls and no other values besides 'True' and 'False' I will convert these values to strings and then move on to the next column.

In [107]:
# Converting the boolean values to strings
orders_df['SUB_IS_ACTIVE_FLAG'] = orders_df['SUB_IS_ACTIVE_FLAG'].map({True:'True',False:'False'})

In [108]:
# Verifying that the changed happened
orders_df['SUB_IS_ACTIVE_FLAG'].apply(type).value_counts()

<class 'str'>    13506
Name: SUB_IS_ACTIVE_FLAG, dtype: int64

## Working on the SUB_PLAN column

In [109]:
# Looking for Nulls in the TOTAL_COST column
sum(orders_df['SUB_PLAN'].isna())

3609

This column refers to the subscription plan a merchant is part of, so there are cases where the merchant didn't subscribe to any plan and that's the reason for Nulls. I want to eyeball this values a bit further before moving on.

In [110]:
# Making the change
orders_df['SUB_PLAN'] = orders_df['SUB_PLAN'].fillna('')

# Verifying that the changed happened
sum(orders_df['SUB_PLAN'].isna())

0

In [111]:
# Getting a glance of the data in this column
orders_df['SUB_PLAN'].value_counts()

business_account_3    8985
                      3609
Plan 4                 790
business_account_1     101
business_account_2      21
Name: SUB_PLAN, dtype: int64

## Working on the SHIPMENT_CARRIER column

In [112]:
# Looking for Nulls in the TOTAL_COST column
sum(orders_df['SHIPMENT_CARRIER'].isna())

1673

This column can also have NULL values because some orders have not been shipped yet or were cancelled. However, I'd like to eyeball the data before moving on.

In [113]:
# Making the change
orders_df['SHIPMENT_CARRIER'] = orders_df['SHIPMENT_CARRIER'].fillna('')

# Verifying that the changed happened
sum(orders_df['SHIPMENT_CARRIER'].isna())

0

In [114]:
# Getting a glance of the data in this column
orders_df['SHIPMENT_CARRIER'].value_counts()

USPS                    9588
                        1673
UPS_MAIL_INNOVATIONS     626
DHL                      388
GLOBEGISTICS             292
CANADA_POST              209
UPS                      172
ASENDIA                  157
ROYAL_MAIL               121
unknown                  115
DPD                       74
DHL_GERMANY               67
ups_                       8
RAYAL_MAIL                 7
DHL_EXPRESS                6
FEDEX                      2
OSM_WORLDWIDE              1
Name: SHIPMENT_CARRIER, dtype: int64

In [115]:
# Based on the previous result, I'd like to know to number of unique values for this column
orders_df['SHIPMENT_CARRIER'].nunique()

17

## Working on the SHIPMENT_DELIVERD_DT column

In [116]:
# Looking for Nulls in the TOTAL_COST column
sum(orders_df['SHIPMENT_DELIVERD_DT'].isna())

2594

Just like in the previous column, I can have Nulls because some orders have not been shipped yet or were cancelled. I don't really need to eyeball more of this datetime column.

In [117]:
# Making the change
orders_df['SHIPMENT_DELIVERD_DT'] = orders_df['SHIPMENT_DELIVERD_DT'].fillna('1900-01-01 00:00:01')

# Verifying that the changed happened
sum(orders_df['SHIPMENT_DELIVERD_DT'].isna())

0

I **finished the data cleaning** of these two dataframes.

# Inserting the clean data from Python to SQL Server

In [118]:
# I need to import a library that will connect Python to SQL Server
import pyodbc

I'll create the parameters needed for the connection, such as the server name and the database and table I previously created in SQL Server to store the records. In order **to insert the records from python I'll use a for loop**, because it will iterate
through all the records of the df automatically and one by one without having to deal with the restriction of only insert 1000 rows at a time.

I'll start transfering the data from the line_items_df.

## Inserting the line_items_df to SQL Server

In [119]:
server = 'DESKTOP-GVTKV9D'
db = 'Printify'
connector_string = 'DRIVER={SQL Server};SERVER='+server+'; DATABASE='+db+'; TRUSTED_CONNECTION=yes'
for i in range(len(line_items_df.index)):
    with pyodbc.connect(connector_string) as cnxn:
        cursor = cnxn.cursor()
        command = "INSERT INTO line_items values" + str(line_items_df.to_records(index = False)[i])
        cursor.execute(command)

In [120]:
# I'll also export the dataframe to a csv file as a backup
line_items_df.to_csv('line_items.csv',index = False)

## Inserting the orders_df to SQL Server

In [121]:
server = 'DESKTOP-GVTKV9D'
db = 'Printify'
connector_string = 'DRIVER={SQL Server};SERVER='+server+'; DATABASE='+db+'; TRUSTED_CONNECTION=yes'
for i in range(len(orders_df.index)):
    with pyodbc.connect(connector_string) as cnxn:
        cursor = cnxn.cursor()
        command = "INSERT INTO orders values" + str(orders_df.to_records(index = False)[i])
        cursor.execute(command)

In [122]:
# I'll also export the dataframe to a csv file as a backup
orders_df.to_csv('Orders.csv',index = False)